File: LeetCode/Longest-Substring-Without-Repeating-Characters/sol.py
# DSA Tracker | Problem: https://leetcode.com/problems/longest-substring-without-repeating-characters/
# DSA Tracker | Language: Python3
# DSA Tracker | Difficulty: Medium
# DSA Tracker | Topics: Hash Table, String, Sliding Window
# DSA Tracker | Runtime: 191 ms — beats 64.20% of submissions
# DSA Tracker | Memory: 19.9 MB — beats 29.59% of submissions

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        maxLength = 0
        charSet = set()
        left = 0
        
        for right in range(n):
            if s[right] not in charSet:
                charSet.add(s[right])
                maxLength = max(maxLength, right - left + 1)
            else:
                while s[right] in charSet:
                    charSet.remove(s[left])
                    left += 1
                charSet.add(s[right])
        
        return maxLength

File: LeetCode/Longest-Substring-Without-Repeating-Characters/sol.cpp
// DSA Tracker | Problem: https://leetcode.com/problems/longest-substring-without-repeating-characters/
// DSA Tracker | Language: C++
// DSA Tracker | Difficulty: Medium
// DSA Tracker | Topics: Hash Table, String, Sliding Window
// DSA Tracker | Runtime: 51 ms — beats 48.05% of submissions
// DSA Tracker | Memory: 19.1 MB — beats 59.69% of submissions

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int maxLength = 0;
        unordered_map<char, int> charMap;
        int left = 0;
        
        for (int right = 0; right < n; right++) {
            if (charMap.count(s[right]) == 0 || charMap[s[right]] < left) {
                charMap[s[right]] = right;
                maxLength = max(maxLength, right - left + 1);
            } else {
                left = charMap[s[right]] + 1;
                charMap[s[right]] = right;
            }
        }
        
        return maxLength;
    }
};
